rewriter/SourceRewriter.cpp

```diff
@@ -252,8 +278,16 @@ class FnPtrTypes : public RefactoringCallback {
     // Replace the old decl with the new one. Make sure to include any macro
     // expansions in the SourceRange being replaced
     clang::CharSourceRange expansion_range = sm.getExpansionRange(*range);
-    Replacement r{sm, expansion_range, new_decl};
-    auto err = file_replacements[file_name].add(r);
+
+    unsigned offset =
+        sm.getFileOffset(sm.getExpansionLoc(expansion_range.getBegin()));
+    auto char_range =
+        clang::CharSourceRange::getTokenRange(expansion_range.getAsRange());
+    auto orig_decl =
+        clang::Lexer::getSourceText(char_range, sm, ctxt.getLangOpts());
+    unsigned int len = orig_decl.size();
```

Comment:
I think you're better off constructing the old replacement as a temporary, then calling its accessors to pull out canonical offset and length to create a new replacement with the new file path, old offset and length. Can even extract that into a helper function. There's some tricky logic in `Replacement::setFromSourceRange()` that I don't think we want to duplicate.

Reply:
I think that should work and would definitely be preferable to this.